test-project/Assets/EditmodeTests/Subscriptions/InjectionCriteriaTests.cs

```diff
@@ -12,118 +13,161 @@ namespace Improbable.Gdk.EditmodeTests.Subscriptions
+                    Assert.IsNotNull(readerBehaviour.Reader);
+                });
         }
 
         [Test]
         public void Writer_is_not_injected_when_entity_component_is_not_checked_out()
         {
-            createdGameObject = CreateAndLinkGameObjectWithComponent<ExhaustiveSingularWriterBehaviour>(EntityId);
-            var writerBehaviour = createdGameObject.GetComponent<ExhaustiveSingularWriterBehaviour>();
-
-            Assert.IsFalse(writerBehaviour.enabled);
-            Assert.IsNull(writerBehaviour.Writer);
+            World
+                .Step(world => { world.Connection.CreateEntity(EntityId, BasicEntity()); })
+                .Step(world =>
+                {
+                    var (_, writerBehaviour) = world.CreateGameObject<ExhaustiveSingularWriterBehaviour>(EntityId);
+                    return writerBehaviour;
+                })
+                .Step((world, writerBehaviour) =>
+                {
+                    Assert.IsFalse(writerBehaviour.enabled);
+                    Assert.IsNull(writerBehaviour.Writer);
+                });
         }
 
         [Test]
         public void Writer_is_not_injected_when_entity_component_is_checked_out_and_unauthoritative()
         {
-            createdGameObject = CreateAndLinkGameObjectWithComponent<PositionWriterBehaviour>(EntityId);
-            var writerBehaviour = createdGameObject.GetComponent<PositionWriterBehaviour>();
-
-            Assert.IsFalse(writerBehaviour.enabled);
-            Assert.IsNull(writerBehaviour.Writer);
+            World
+                .Step(world => { world.Connection.CreateEntity(EntityId, BasicEntity()); })
+                .Step(world =>
+                {
+                    var (_, writerBehaviour) = world.CreateGameObject<PositionWriterBehaviour>(EntityId);
+                    return writerBehaviour;
+                })
+                .Step((world, writerBehaviour) =>
+                {
+                    Assert.IsFalse(writerBehaviour.enabled);
+                    Assert.IsNull(writerBehaviour.Writer);
+                });
         }
 
         [Test]
         public void Writer_is_injected_when_entity_component_is_checked_out_and_authoritative()
         {
-            ConnectionHandler.ChangeAuthority(EntityId, Position.ComponentId, Authority.Authoritative);
-            ReceiveSystem.Update();
-
-            createdGameObject = CreateAndLinkGameObjectWithComponent<PositionWriterBehaviour>(EntityId);
-            var writerBehaviour = createdGameObject.GetComponent<PositionWriterBehaviour>();
-
-            Assert.IsTrue(writerBehaviour.enabled);
-            Assert.IsNotNull(writerBehaviour.Writer);
+            World
+                .Step(world =>
+                {
+                    world.Connection.CreateEntity(EntityId, BasicEntity());
+                    world.Connection.ChangeAuthority(EntityId, Position.ComponentId, Authority.Authoritative);
```

Comment:
Do we want to make the component type safe here by using a generic type and the database instead of a uint?

Reply:
Why is this not type-safe currently?